core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -396,7 +396,16 @@ abstract class RDD[T: ClassTag](
    * Return a new RDD containing the distinct elements in this RDD.
    */
   def distinct(numPartitions: Int)(implicit ord: Ordering[T] = null): RDD[T] = withScope {
-    map(x => (x, null)).reduceByKey((x, y) => x, numPartitions).map(_._1)
+    // If the data is already approriately partitioned with a known partitioner we can work locally.
+    def removeDuplicatesInPartition(itr: Iterator[T]): Iterator[T] = {
+      val set = new mutable.HashSet[T]()
+      itr.filter(set.add(_))
```

Comment:
not a big deal, but despite this is really compact and elegant, it adds to the set also the elements which are already there and it is not needed. We can probably check if the key is there and add it only in that case, probably it is a bit faster.

Reply:
So according to HashSet can only contain one instance for each element so we don't need to worry about adding multiple copies of the elements.

Reply:
But we need to compute the hash of the key anyways to check if it exists.

Reply:
yes, the hash should be computed anyway, this is just a nit, I am not sure if the perf gain would be even noticeable, that is why I already gave my LGTM, despite this comment.

Reply:
This is a bad implementation and could OOM. You should reuse reduceByKey.

Reply:
This is a bad implementation and could OOM. You should reuse reduceByKey.

Reply:
So to reuse `reduceByKey` I'd write a custom partitioner which uses the existing partioner as it's base but takes in the combined key type as input and drop it down to the original key.

Sound right?